Replace Network-interfaces.target

groxxda opened the pull request:
###### Motivation for this change

**This is ready for review and test**

Systemd upstream provides 3 targets that should be used in services to allow ordering and dependencies for networking (network, network-pre, network-online)
Nixos adds at least network-interfaces.target to the list where it is not clear what it's purpose is.
To make network setup easier to understand for module developers this is an effor to get rid of the superfluous target.
This is a follow up on the removal of up-up.target and is not yet ready for merge.
~~I post this merely for early review and discussion (Hijacking an unrelated or was not a good idea in the first place)~~

~~Most~~ commits should be safe to merge ~~(getting closer to upstream stuff)~~

**Especially the changes to the scripted network interface setup should be reviewed with care** even though all networking tests passed with them.

~~See also https://github.com/joachifm/nixpkgs/tree/network-interfaces~~

Ping @joachifm 

Next steps:
- [x] Merge joachifm's branch
- [ ] Review all changes, especially the work on network-interfaces-scripted.nix
- [x] Cleanup the mess that accumulated in all these commits
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

groxxda committed:
networking.dhcpcd: use upstream targets

groxxda committed:
treewide: minor format / style / documentation fixes

groxxda committed:
networking.nat: replace network-interfaces.target

We can replace this safely with network-pre because iptables does not
care whether the interfaces exist or not.

groxxda committed:
multi-user.target should not pull network.target

groxxda committed:
network-interfaces-scripted: rework dependencies

joachifm committed:
cassandra service: network-interfaces.target -> network.target

joachifm committed:
influxdb service: network-interfaces.target -> network.target

joachifm committed:
neo4j service: network-interfaces.target -> network.target

joachifm committed:
apache-kafka service: network-interfaces.target -> network.target

joachifm committed:
folding-at-home service: network-interfaces.target -> network.target

joachifm committed:
mesos-master service: network-interfaces.target -> network.target

joachifm committed:
mesos-slave service: network-interfaces.target -> network.target

joachifm committed:
svnserve service: network-interfaces.target -> network.target

joachifm committed:
zookeeper service: network-interfaces.target -> network.target

joachifm committed:
graphite service: network-interfaces.target -> network.target

joachifm committed:
nagios service: network-interfaces.target -> network.target

joachifm committed:
openafs-client service: network-interfaces.target -> network.target

joachifm committed:
bind service: network-interfaces.target -> network.target

joachifm committed:
cjdns service: network-interfaces.target -> network.target

joachifm committed:
oidentd service: network-interfaces.target -> network.target

joachifm committed:
radicale service: network-interfaces.target -> network{,-online}.target

joachifm committed:
tcpcrypt service: remove use of network-interfaces.target

joachifm committed:
tinc service: remove use of network-interfaces.target

joachifm committed:
wpa-supplicant service: remove use of network-interfaces.target

joachifm committed:
xinetd service: remove use of network-interfaces.target

joachifm committed:
zerobin service: remove use of network-interfaces.target

joachifm committed:
chronos service: remove use of network-interfaces.target

joachifm committed:
marathon service: remove use of network-interfaces.target

joachifm committed:
elasticsearch service: remove use of network-interfaces.target

joachifm committed:
kibana service: remove use of network-interfaces.target

joachifm committed:
oauth2_proxy: remove use of network-interfaces.target

joachifm committed:
peerflix: remove use of network-interfaces.target

joachifm committed:
tomcat: remove use of network-interfaces.target

groxxda commented:
I just rebased against current master with the changes from #18319 
Here's a merge conflict that looks relevant:
@fpletz changed `dhcpcd.service` in https://github.com/NixOS/nixpkgs/commit/c58654e2b78e7d3ae97ba96c80ceb3f96236e463 to `Before = network.target` (like [raspberrypi](https://github.com/raspberrypi/dhcpcd5-6.9.4/blob/master/debian/dhcpcd5.dhcpcd.service) does) while I added `After = network.target` (like [coreos](https://github.com/coreos/portage-stable/blob/master/net-misc/dhcpcd/files/dhcpcd.service) [chromiumos](https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay/+/master/net-misc/dhcpcd/files/dhcpcd.service) [archlinux](https://aur.archlinux.org/cgit/aur.git/tree/dhcpcd.service?h=dhcpcd-trunk) do).

joachifm commented:
Hm, network.target is supposed to mean that networking functionality is "available". To me, the intuitive choice is to order dhcpcd after network.target.

Ordering before/after causes dhcpcd to be stopped after/before network.target on shutdown, not sure if that makes any practical difference to us, however.

groxxda committed:
wicd: get closer to upstream service definition

taken from
http://bazaar.launchpad.net/~wicd-devel/wicd/experimental/view/head:/other/wicd.service

groxxda committed:
softether: improve service dependencies

groxxda committed:
monitoring.monit: get closer to upstream service definition

taken from
https://bitbucket.org/tildeslash/monit/src/e02247e0480b19be88e6a908886da83bbb1bc6a3/system/startup/monit.service.in?at=master

groxxda committed:
misc.etcd: get closer to upstream service definition

taken from
https://github.com/coreos/etcd/blob/master/contrib/systemd/etcd.service

I intentionally kept "After = network.target" because I think it's
missing upstream (https://github.com/coreos/etcd/pull/6388)

groxxda committed:
cluster.kubernetes: fix service ordering

Requires does NOT imply After, so I added the missing ordering.

groxxda committed:
gpve service: network-interfaces.target -> network.target

groxxda committed:
openvpn service: network-interfaces.target -> network.target

groxxda committed:
lshd service: remove use of network-interfaces.target

groxxda committed:
systemd-networkd.service: remove network-interfaces.target ordering

fpletz commented:
`Before` or `After` `network.target` for dhcpcd are both fine by me. It doesn't really matter because dhcpcd detects if networking and interfaces are or become available and acts accordingly.

fpletz commented:
The changes to scripted network interfaces look fine to me.

groxxda committed:
network-interfaces.target: add deprecation notice

groxxda committed:
tests.networking: remove network-interfaces.target

joachifm commented:
@groxxda how are you feeling about this? Is it ready from your pov? I think it's been up long enough that we can begin thinking about integrating it.

groxxda commented:
@joachifm I would be happy to see this merged. The closer we are to the next release the higher the risk will be. 
Note that I mainly tested with networkd (scripted networking on my laptop doesn't really count). I hope the scripted networking doesn't break..
Would be awesome if someone would test this first on a machine that uses scripted networking and has an advanced network configuration (@fpletz maybe take a shot at one of those cosmodromes? :rocket:  :see_no_evil: )

joachifm committed:
network-interfaces-scripted: DefaultDependencies is a unitConfig

Ref: 072c1dcc4a319b80ffd4a41165cebacdbb17275a

joachifm commented:
I'm working on integrating this today, going to do some testing and stuff just to be sure we're not breaking the world :)

groxxda commented:
Thanks @joachifm :tada: Let me know if there's something I can do. I'm happy to help out to get these changes merged

joachifm commented:
@groxxda could you run a few of the nixos tests that might be affected? In particular, stuff like misc, etcd, openssh, firewall, and any other that looks important to you.

joachifm commented:
I'm having a problem actually getting tests to run locally atm, stuff is just timing out on me for no obvious reason (with and without this PR).

groxxda commented:
@joachifm I'll see if I can run some tests. Last time qemu segfaulted sporadically giving false-negatives..

By the way you should be able to push directly to my branch (`Allow edits from maintainers` is checked). Feel free to make use of it :+1:

joachifm commented:
I think we'll have to accept that some regressions may be introduced by this, so it's not a complete blocker seeing as nobody has brought up any fundamental objections thus far, but it'd be nice to not break everything at once :)

groxxda commented:
Tests:
misc: 17 out of 17 tests succeeded
etcd: 2 out of 2 tests succeeded
firewall: finishes
openssh: 2 out of 2 tests succeeded
initrdNetwork: finishes

groxxda commented:
`nix-build '<nixpkgs/nixos/tests/networking.nix>' --arg networkd false` finishes

Mic92 commented on nixos/modules/services/monitoring/monit.nix:
```diff
@@ -36,11 +36,16 @@ in
     systemd.services.monit = {
-      description = "Monit system watcher";
-      after = [ "network-interfaces.target" ];
+      description = "Pro-active monitoring utility for unix systems";
```
While this is a good description, what monit does it obscures the name if on the boot screen "Pro-active monitoring utility for unix systems started"  appears.

groxxda replied:
That's right. I simply took it from the upstream file at https://bitbucket.org/tildeslash/monit/src/e02247e0480b19be88e6a908886da83bbb1bc6a3/system/startup/monit.service.in?at=master
Should we revert?

joachifm replied:
Sorry, I don't understand the problem? Is it that it doesn't mention the name of the unit being started?

Mic92 replied:
yes.

joachifm commented:
Alright, let's do it.

groxxda commented:
Thanks again for reviewing and pushing this forward! :beers:

joachifm commented:
A potential regression: https://hydra.nixos.org/build/41566104

groxxda commented:
@joachifm More likely one of these _times out because hydra load is high_ failures..
Looks similar to https://hydra.nixos.org/build/41444005

edolstra commented:
@groxxda 

This PR changed several units to be pulled in by `network-setup.service` rather than a target (`network-interfaces.target`). This is wrong. On a switch to a new configuration. `network-setup.service` is not necessarily restarted, so any new dependencies will not be started. Target units OTOH are treated specially by switch-to-configuration and are started even when they are already running to ensure new dependencies are activated

For example, if I add the following to my configuration:
```
  networking.interfaces.enp0s31f6.ip4 =
    [ { address = "192.168.1.9";
        prefixLength = 24;
      }
    ];
```
than this will have no effect until I reboot my system or explicitly start `network-addresses-enp0s31f6.service`.

Probably these units should be made wanted by one of the networking targets.

edolstra commented:
I'm also worried about the use of `partOf = [ "network-setup.service" ]` in various units, which means that stopping network-setup.service during switch-to-configuration will kill all statically configured networking interfaces. (See https://github.com/NixOS/nixops/issues/640, though there the dependent units are also stopped directly.)

groxxda commented:
Replacing `network-setup.services` by `network-setup.target` should fix the problems, right?

edolstra commented:
Yeah that sounds good.
